pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs

```diff
@@ -277,19 +278,56 @@ data LegacyBootEvent
     | Dawn Dawn
   deriving (Eq, Show)
 
-data ArrowKey = D | L | R | U
-  deriving (Eq, Ord, Show)
-
-data Belt
-    = Aro ArrowKey
+data Bolt
+    = Key Char
+    | Aro ArrowKey
     | Bac ()
-    | Ctl Cord
     | Del ()
-    | Met Cord
+    | Hit Word64 Word64
     | Ret ()
+  deriving (Eq, Ord, Show)
+
+data Belt
+    = Bol Bolt
+    | Mod Modifier Bolt
     | Txt Tour
   deriving (Eq, Ord, Show)
 
+data ArrowKey = D | L | R | U
+  deriving (Eq, Ord, Show)
+
+data Modifier = Ctl | Met | Hyp
+  deriving (Eq, Ord, Show)
+
+--NOTE  required to get the above declarations into reify's type environment
+--      see also ghc/ghc#9813
+$(pure [])
+
+instance FromNoun Bolt where
+  parseNoun = \case
+    A c -> pure $ Key $ C.chr $ fromIntegral c
+    n   -> $(deriveFromNounFunc ''Bolt) n
+
+instance FromNoun Belt where
+  parseNoun = \case
+    A c       -> pure $ Bol $ Key $ C.chr $ fromIntegral c
+    n         -> runParser ($(deriveFromNounFunc ''Bolt) n) [] belt bolt
+                   where
+                    belt p m = $(deriveFromNounFunc ''Belt) n
+                    bolt !b  = pure $ Bol b
```

Comment:
Suggested code:

- parse the noun coming in into a Maybe Bolt
- if it's just (use `maybe`, possibly), wrap in Bol
- otherwise, call the derived Belt parser

Reply:
Actually, you ought to be able to accomplish the above using the Alternative instance of parsers. It should look something like

```
parseNoun n = Bol <$> parseNoun n <|> $(deriveFromNounFunc ''Belt) n
```

which also very clearly and cleanly expresses what is going on

Reply:
> which also very clearly and cleanly expresses what is going on

![image](https://user-images.githubusercontent.com/3829764/114733893-03155a00-9d44-11eb-80e0-1a1d561127eb.png)